### PR TITLE
Specify groups

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## Content
- group endpoint
- interlinking of groups and users (incl. endpoints)
- groups are also valid instead of users in many places
## Changes to current implementation
- users `mail` property is now called `email`
- the schema of a form now features multiple types per field instead of one

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://feature-specify-groups.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

File: doc/apiv3-documentation.apib [hunk 1]
```diff
@@ -886,6 +886,120 @@ This is an example of how a form might look like. Note that this endpoint does n
                 "message": "The resource you are about to edit was changed in the meantime."
             }
 
+# Group Groups
+
+## Linked Properties:
+|  Link               | Description                              | Type          | Nullable | Supported operations |
+|:-------------------:|----------------------------------------- | ------------- | -------- | -------------------- |
+| self                | This group                               | Group         |          | READ                 |
+| members             | This group                               | Users         |          | READ                 |
```
Comment (ulferts) [inline doc/apiv3-documentation.apib]: ahm...

**Comment by ulferts** [comment]
With the exception of the TODO (specifying the valid places groups can be used in) you already mentioned, this is fine with me.

File: doc/apiv3-documentation.apib (hunk 1 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: I was just testing you

:crocodile:

**Comment by NobodysNightmare** [comment]
I also had to change the format of our schema.
Instead of specifying the (single) `type` of a field, the schema now specifies the _possible_ `types` of a field, which happens to be a single type for most fields.

So what we have now is

``` JSON
"types": [ "String" ]
```

for something like a subject, but

``` JSON
"types": [ "User", "Group" ]
```

for assignees and responsibles.

This is basically the consequence of not advertising inheritance. If we _had_ inheritance we could stick with a single type (say `Principal`), but now that User and Group are unrelated, we have to go for the "multi type way".

**Comment by myabc** [comment]
Rebase me!

**Comment by floriank** [comment]
@NobodysNightmare: What @myabc said!

**Comment by basiszwo** [comment]
Closing this as it obviously wasn't important enough to be finished.

**Comment by NobodysNightmare** [comment]
> ... it obviously wasn't important enough to be finished.

Not my take on the story, but I won't finish it either...

**Pull request closed without merging**